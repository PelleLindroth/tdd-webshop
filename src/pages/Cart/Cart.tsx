import { useEffect, useState } from 'react'
import { IProduct, productsDb } from '../../mocks/products'
interface CartProps {
  cart: IProduct[]
  setCart: React.Dispatch<React.SetStateAction<IProduct[]>>
}

interface CartWithAmount extends IProduct {
  amount?: number
}

const Cart: React.FC<CartProps> = (props: CartProps) => {
  const { cart, setCart } = props
  const [grandTotal, setGrandTotal] = useState(0)

  const [cartObject, setCartObject] = useState({})

  useEffect(() => {
    setCartObject(
      cart.reduce((acc: any, currItem) => {
        if (acc[currItem.id]) {
          acc[currItem.id].amount++
          acc[currItem.id].price += currItem.price
        } else {
          acc[currItem.id] = { ...currItem, amount: 1 }
        }
        return acc
      }, {})
    )
    setGrandTotal(cart.reduce((acc: any, currItem) => acc + currItem.price, 0))
  }, [cart])

  const handleIncrement = (id: string) => {
    const product = productsDb.find((item) => item.id === id)
    setCart([...cart, product!])
  }

  const handleDecrement = (id: string) => {
    const cartItemIndex = cart.findIndex((item) => item.id === id)
    cart.splice(cartItemIndex, 1)
    setCart([...cart])
  }

  const handleDelete = (id: string) => {
    const updatedCart = cart.filter((item) => item.id !== id)
    setCart([...updatedCart])
  }

  const cartArr: CartWithAmount[] = Object.values(cartObject)

  return (
    <>
      <h1>Cart</h1>
      {cart.length > 0 ? (
        <ul>
          {cartArr.map((el) => {
            return (
              <li key={el.id}>
                <CartRow cartItem={el} />
                <button
                  data-testid={`increment-${el.name}`}
                  onClick={() => handleIncrement(el.id)}
                >
                  +
                </button>
                <button
                  data-testid={`decrement-${el.name}`}
                  onClick={() => handleDecrement(el.id)}
                >
                  -
                </button>
                <button
                  data-testid={`delete-${el.name}`}
                  onClick={() => handleDelete(el.id)}
                >
                  Delete
                </button>
              </li>
            )
          })}
        </ul>
      ) : (
        <h2>Cart is empty. Buy something!</h2>
      )}
      <h2>Grand total: SEK {grandTotal}</h2>
    </>
  )
}

export default Cart

const CartRow = (props: { cartItem: CartWithAmount }) => {
  const { cartItem } = props
  return (
    <div data-testid={`${cartItem.name}-item`}>
      <h2>{cartItem.name}</h2>
      <p data-testid={`${cartItem.name}-amount`}>Amount: {cartItem.amount}</p>
      <p data-testid={`${cartItem.name}-price`}>SEK {cartItem.price}</p>
    </div>
  )
}

// [
//   { name: 'Bike', id: '234', price: 3456 },
//   { id: '4', name: 'desk', price: 1200 },
//   { id: '2', name: 'A table', price: 1400 },
//   { id: '3', name: 'A car', price: 150000 },
//   { id: '4', name: 'desk', price: 1200 },
//   { id: '2', name: 'A table', price: 1400 },
// ]
