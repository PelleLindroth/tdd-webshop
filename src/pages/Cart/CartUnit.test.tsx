import { shallow } from 'enzyme'
import { render, screen } from '@testing-library/react'
import Cart from './Cart'
import userEvent from '@testing-library/user-event'

const mockCart = [
  { id: '234', name: 'bike', price: 3456 },
  { id: '4', name: 'desk', price: 1200 },
  { id: '2', name: 'table', price: 1400 },
  { id: '3', name: 'car', price: 150000 },
  { id: '4', name: 'desk', price: 1200 },
  { id: '2', name: 'table', price: 1400 },
]

describe('Cart', () => {
  it('shallow renders Cart component correctly', () => {
    const wrapper = shallow(<Cart setCart={jest.fn()} cart={[]} />)

    expect(wrapper).toMatchSnapshot()
  })
  it('shows all products in Cart', () => {
    render(<Cart cart={mockCart} setCart={jest.fn()} />)

    const products = screen.getAllByRole('listitem')

    expect(products).toHaveLength(4)
  })
  it('shows correct amount of desks', () => {
    render(<Cart cart={mockCart} setCart={jest.fn()} />)

    const deskAmount = screen.getByTestId('desk-amount')

    expect(deskAmount).toHaveTextContent('2')
  })
  it('shows correct price for 1 car', () => {
    render(<Cart cart={mockCart} setCart={jest.fn()} />)

    const carPrice = screen.getByTestId('car-price')

    expect(carPrice).toHaveTextContent('150000')
  })
  it('shows correct price for 2 desks', () => {
    render(<Cart cart={mockCart} setCart={jest.fn()} />)

    const deskPrice = screen.getByTestId('desk-price')

    expect(deskPrice).toHaveTextContent('2400')
  })
  it('calls setCart once when + button is clicked', async () => {
    const setCartSpy = jest.fn()
    render(<Cart cart={mockCart} setCart={setCartSpy} />)

    const incrementCarButton = screen.getByTestId('increment-car')

    userEvent.click(incrementCarButton)

    expect(setCartSpy).toHaveBeenCalledTimes(1)
    expect(setCartSpy).toHaveBeenCalledWith([
      ...mockCart,
      { id: '3', name: 'car', price: 150000 },
    ])
  })
  it('calls setCart once when - button is clicked', async () => {
    const setCartSpy = jest.fn()
    render(<Cart cart={mockCart} setCart={setCartSpy} />)

    const decrementDeskButton = screen.getByTestId('decrement-desk')

    userEvent.click(decrementDeskButton)

    expect(setCartSpy).toHaveBeenCalledTimes(1)
    expect(setCartSpy).toHaveBeenCalledWith([...mockCart])
  })
  it('shows a message if cart is empty', () => {
    render(<Cart cart={[]} setCart={jest.fn()} />)

    const message = screen.getByText(/empty/i)

    expect(message).toBeInTheDocument()
  })
  it('shows no message if cart contains products', () => {
    render(<Cart cart={mockCart} setCart={jest.fn()} />)

    const message = screen.queryByText(/empty/i)

    expect(message).toBeNull()
  })
  it('shows grand total correctly', () => {
    render(<Cart cart={mockCart} setCart={jest.fn()} />)

    const grandTotal = screen.getByText(/grand total/i)

    expect(grandTotal).toHaveTextContent('Grand total: SEK 157456')
  })
})
