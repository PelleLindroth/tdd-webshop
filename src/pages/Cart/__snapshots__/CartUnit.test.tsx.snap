// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart unit tests shallow renders Cart component correctly 1`] = `
<Fragment>
  <h1>
    Cart
  </h1>
  <h2>
    Cart is empty. Buy something!
  </h2>
  <h2>
    Grand total: SEK 
    0
  </h2>
</Fragment>
`;
